core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java

```diff
@@ -462,36 +462,39 @@ protected void regularInvalidationFinished(Object[] keys, int[] topologyIds, lon
                });
             }
          }
-      } while (removedKeys != this.removedKeys);
-      if (removedKeys.size() > invalidationBatchSize || (force && !removedKeys.isEmpty())) {
-         tryRemovedInvalidations(removedKeys);
+         needsSend = removedKeys.size() > invalidationBatchSize || (force && !removedKeys.isEmpty());
+      } finally {
+         readLock.unlock();
+      }
+      if (needsSend) {
+         tryRemovedInvalidations();
       }
    }
 
-   private void tryRemovedInvalidations(ConcurrentMap<K, InvalidationInfo> removedKeys) {
-      if (removedKeysSwapper.compareAndSet(this, removedKeys, new ConcurrentHashMap<>(invalidationBatchSize))) {
-         final ConcurrentMap<K, InvalidationInfo> rk = removedKeys;
-         executorService.execute(() -> {
-            int numKeys = rk.size();
-            Object[] keys = new Object[numKeys];
-            int[] topologyIds = new int[numKeys];
-            long[] versions = new long[numKeys];
-            int i = 0;
-            for (Map.Entry<K, InvalidationInfo> entry : rk.entrySet()) {
-               keys[i] = entry.getKey();
-               topologyIds[i] = entry.getValue().topologyId;
-               versions[i] = entry.getValue().version;
-               if (++i > numKeys) {
-                  numKeys = rk.size();
-                  keys = Arrays.copyOf(keys, numKeys);
-                  topologyIds = Arrays.copyOf(topologyIds, numKeys);
-                  versions = Arrays.copyOf(versions, numKeys);
-               }
-            }
-            InvalidateVersionsCommand removeCommand = commandsFactory.buildInvalidateVersionsCommand(-1, keys, topologyIds, versions, true);
-            sendRemoveInvalidations(removeCommand);
-         });
+   private void tryRemovedInvalidations() {
+      final ConcurrentMap<K, InvalidationInfo> removedKeys;
+      Lock writeLock = removedKeysLock.writeLock();
+      writeLock.lock();
+      try {
+         removedKeys = this.removedKeys;
+         this.removedKeys = new ConcurrentHashMap<>(invalidationBatchSize);
+      } finally {
+         writeLock.unlock();
       }
+      executorService.execute(() -> {
```

Comment:
Maybe optionally invoke this only if `removedKeys.size() > 0` and then this method can return if it is removing invalidations? Is it possible to check `isEmpty` in `startFlush` and see it is empty and then get an insert afterwards?

Reply:
Do you mean moving the checks from lines 203 and 469 here? Yes, I can do that..

Reply:
Yeah this way you can't add to the ConcurrentMap in between since they are swapped. This could prevent possible missed `removedKeys`, but I am not sure if that was possible before.

Reply:
@wburns Not sure if I follow: if you add to `removedKeys` after the `.isEmpty` check all that happens is that the invalidations are sent a bit later. Since you swap with a write-lock and add with a read-lock, the situation where you'd add a key to removedKeys and this would be dropped without sending is not possible, IMO.

Reply:
Yeah looking closer the method I was more worried about was `startFlush` which I didn't realize was only for testing. And as you mentioned it doesn't replace the map either, so it should be fine.